core/tests/protractor_utils/ExplorationEditorImprovementsTab.js

```diff
@@ -25,46 +25,181 @@ var ruleTemplates = require(
-  var closeAnswerDetailsButton = element(
-    by.css('.protractor-test-close-answer-details'));
+  const cardsContainer = $('.protractor-test-cards');
+  const closeModalButton = $('.protractor-test-close-modal-button');
 
-  var _getAnswerDetailsCardStateName = function() {
-    return answerDetailsCardStateName.getText();
+  const answerDetails = $('.protractor-test-answer-details');
+  const answerInfoCount = $('.protractor-test-answer-info-count');
+
+  const threadMessagesContainer = $('.protractor-test-thread-messages');
+  const responseTextarea = $('.protractor-test-response-textarea');
+  const responseStatusSelect = $('.protractor-test-response-status-select');
+  const responseSendButton = $('.protractor-test-response-send-button');
+  const reviewSuggestionButton = $('.protractor-test-review-suggestion-button');
+
+  const suggestionRows = $$('.protractor-test-oppia-feedback-tab-row');
+  const acceptSuggestionButton = $(
+    '.protractor-test-exploration-accept-suggestion-btn');
+  var rejectSuggestionButton =
+    $('.protractor-test-exploration-reject-suggestion-btn');
+  var suggestionReviewMessageInput =
+    $('.protractor-test-suggestion-review-message');
+  var suggestionCommitMessageInput =
+    $('.protractor-test-suggestion-commit-message');
+
+  const threadMessageLocator =
+    by.css('.protractor-test-thread-message-body');
+  const threadMessageAuthorLocator =
+    by.css('.protractor-test-thread-message-author');
+  const threadMessageStatusChangeLocator =
+    by.css('.protractor-test-thread-message-status-change');
+
+  const actionButtonLocator = by.css('.protractor-test-action-button');
+  const cardLocator = by.css('.protractor-test-card');
+  const cardBodyLocator = by.css('.protractor-test-card-body');
+  const cardStatusLocator = by.css('.protractor-test-card-status');
+  const cardTitleLocator = by.css('.protractor-test-card-title');
+  const stateNameLocator = by.css('.protractor-test-card-state-name');
+
+  const suggestionRowClassName = '.protractor-test-oppia-feedback-tab-row';
+
+  var _cardHasStateName = function(card, stateName) {
+    return card.element(stateNameLocator).getText()
+      .then(text => text === stateName);
+  };
+
+  var _cardHasTitle = function(card, titleSubstring) {
+    return card.element(cardTitleLocator).getText()
+      .then(text => text.includes(titleSubstring));
+  };
+
+  var _cardHasContent = function(card, content) {
+    return card.element(cardBodyLocator).getText()
+      .then(text => text.includes(content));
   };
 
-  var _getAnswerInfoCount = function() {
-    return answerInfoCount.getText();
+  var _cardHasType = function(card, cardType) {
+    return card.getAttribute('class')
+      .then(cardCssClass => cardCssClass.includes(cardType));
   };
 
-  var _getAnswerDetails = function() {
-    return answerDetails.getText();
+  /**
+   * @typedef CardMatchOptions
+   * @property {string} card_type - the type of the card.
```

Comment:
Seems like card_title is a possibility as well.

I'm not really a fan of the dicts with different keys, TBH. I would suggest passing 4 arguments explicitly, using an explicit null for arguments which we don't need a predicate for.

Reply:
I disagree -- I think explicit args are easier to maintain than dict blobs, but perhaps harder to code in-the-moment. I would strongly discourage using dicts in this way, since they are too amorphous, and this is something we are trying to get away from in the codebase as a whole (especially as things become more typed).

So I would still request changing it; I'm also happy to discuss this more when we meet tomorrow, if you like.

Reply:
OK, I've done a bit of a compromise. How's that?